src/core/common/error.hpp

```diff
@@ -0,0 +1,105 @@
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file defines the errors used by OpenThread core.
+ */
+
+#ifndef ERROR_HPP_
+#define ERROR_HPP_
+
+#include "openthread-core-config.h"
+
+#include <openthread/error.h>
+
+namespace ot {
+
+/*
+ * The `OT_ERROR` enumeration values are re-defined using `kError` style names.
+ * See `openthread/error.h` for list of errors.
+ *
+ */
+#define kErrorNone OT_ERROR_NONE
+#define kErrorFailed OT_ERROR_FAILED
+#define kErrorDrop OT_ERROR_DROP
+#define kErrorNoBufs OT_ERROR_NO_BUFS
+#define kErrorNoRoute OT_ERROR_NO_ROUTE
+#define kErrorBusy OT_ERROR_BUSY
+#define kErrorParse OT_ERROR_PARSE
+#define kErrorInvalidArgs OT_ERROR_INVALID_ARGS
+#define kErrorSecurity OT_ERROR_SECURITY
+#define kErrorAddressQuery OT_ERROR_ADDRESS_QUERY
+#define kErrorNoAddress OT_ERROR_NO_ADDRESS
+#define kErrorAbort OT_ERROR_ABORT
+#define kErrorNotImplemented OT_ERROR_NOT_IMPLEMENTED
+#define kErrorInvalidState OT_ERROR_INVALID_STATE
+#define kErrorNoAck OT_ERROR_NO_ACK
+#define kErrorChannelAccessFailure OT_ERROR_CHANNEL_ACCESS_FAILURE
+#define kErrorDetached OT_ERROR_DETACHED
+#define kErrorFcs OT_ERROR_FCS
+#define kErrorNoFrameReceived OT_ERROR_NO_FRAME_RECEIVED
+#define kErrorUnknownNeighbor OT_ERROR_UNKNOWN_NEIGHBOR
+#define kErrorInvalidSourceAddress OT_ERROR_INVALID_SOURCE_ADDRESS
+#define kErrorAddressFiltered OT_ERROR_ADDRESS_FILTERED
+#define kErrorDestinationAddressFiltered OT_ERROR_DESTINATION_ADDRESS_FILTERED
+#define kErrorNotFound OT_ERROR_NOT_FOUND
+#define kErrorAlready OT_ERROR_ALREADY
+#define kErrorIp6AddressCreationFailure OT_ERROR_IP6_ADDRESS_CREATION_FAILURE
+#define kErrorNotCapable OT_ERROR_NOT_CAPABLE
+#define kErrorResponseTimeout OT_ERROR_RESPONSE_TIMEOUT
+#define kErrorDuplicated OT_ERROR_DUPLICATED
+#define kErrorReassemblyTimeout OT_ERROR_REASSEMBLY_TIMEOUT
+#define kErrorNotTmf OT_ERROR_NOT_TMF
+#define kErrorNotLowpanDataFrame OT_ERROR_NOT_LOWPAN_DATA_FRAME
+#define kErrorLinkMarginLow OT_ERROR_LINK_MARGIN_LOW
+#define kErrorInvalidCommand OT_ERROR_INVALID_COMMAND
+#define kErrorPending OT_ERROR_PENDING
+#define kErrorGeneric OT_ERROR_GENERIC
+
+#define kNumErrors OT_NUM_ERRORS
+
+/**
+ * This type represents error codes used by OpenThread core modules.
+ *
+ */
+typedef otError Error;
+
+/**
+ * This function converts an `Error` into a string.
+ *
+ * @param[in]  aError     An error.
+ *
+ * @returns  A string representation of @p aError.
+ *
+ */
+const char *ErrorToString(Error aError);
+
```

Comment:
```suggestion
typedef otError Error;

OT_TOOL_PACKED_BEGIN
class Error {
public:
  /**
   * This function converts an `Error` into a string.
   *
   * @param[in]  aError     An error.
   *
   * @returns  A string representation of @p aError.
   *
   */
    const char *ToString(void);
    
    otError mError;
} OT_TOOL_PACKED_END;
```

let's make `Error` a class ?

This would allow us adding more useful methods, e.g. `error.Ignore()` instead of `IgnoreError(error)`.

Reply:
I would also like to make `Error` be a `class` and it was what I went for first.  😃 But realized that unfortunately it adds a lot overhead and increases the code size (I guess the toolchains cannot optimize `Error` as a class (containing an `enum`) as efficiently as a simple `enum`).

Here is the code size impact:
```
 text	   data	    bss	    dec	    hex	filename
 400376	    420	  76308	 477104	  747b0	output/nrf52840/bin/ot-ncp-ftd  (`Error` is a typedef)
 403064	    420	  76308	 479792	  75230	output/nrf52840/bin/ot-ncp-ftd  (`Error` is a class)  + 2.68K

 text	   data	    bss	    dec	    hex	filename
 329912	    420	  67700	 398032	  612d0	output/nrf52840/bin/ot-ncp-mtd  (`Error` is a typedef)
 331608	    420	  67700	 399728	  61970	output/nrf52840/bin/ot-ncp-mtd  (`Error` is a class) + 1.68K
```

Reply:
Thanks for the information. SGTM.